src/compaction_filter.h

```diff
@@ -0,0 +1,111 @@
+#pragma once
+
+#include "db_impl.h"
+#include "rocksdb/compaction_filter.h"
+
+#include <utility>
+
+namespace rocksdb {
+namespace titandb {
+
+class TitanCompactionFilter final : public CompactionFilter {
+public:
+  explicit TitanCompactionFilter(const TitanDBImpl *db,
+                                 const CompactionFilter *origin, uint32_t cf_id)
+      : titan_db_impl_(db), origin_filter_(origin), column_family_id_(cf_id) {}
+
+  const char *Name() const override { return "TitanCompactionFilter"; }
+
+  Decision FilterV2(int level, const Slice &key, ValueType value_type,
+                    const Slice &value, std::string *new_value,
+                    std::string *skip_until) const override {
+    if (value_type != kBlobIndex) {
+      return origin_filter_->FilterV2(level, key, value_type, value, new_value,
+                                      skip_until);
+    }
+
+    BlobIndex blob_index;
+    Slice origin_value(value.data());
+    Status s = blob_index.DecodeFrom(&origin_value);
+    if (!s.ok()) {
+      // Unable to decode blob index. Keeping the value.
+      return Decision::kKeep;
+    }
+    if (BlobIndex::IsDeletionMarker(blob_index)) {
+      return Decision::kRemove;
+    }
+
+    BlobRecord record;
+    PinnableSlice buffer;
+
+    titan_db_impl_->mutex_.Lock();
+    auto storage =
+        titan_db_impl_->blob_file_set_->GetBlobStorage(column_family_id_)
+            .lock();
+    titan_db_impl_->mutex_.Unlock();
+
+    if (storage) {
+      ReadOptions read_options;
+      s = storage->Get(read_options, blob_index, &record, &buffer);
+    } else {
+      // Column family not found, remove the value.
+      return Decision::kRemove;
+    }
+
+    if (s.ok()) {
+      bool value_changed = false;
+      bool rv = origin_filter_->Filter(level, key, record.value, new_value,
```

Comment:
why not use `FilterV2`?

Reply:
这里调用`FilterV2`会产生和上述`value_changed == true`一样的问题，因此选择调用`Filter`。将来rocksdb本身的API可以支持`kBlobIndex`后可改为调用`FilterV2`

Reply:
I still don't understand why `FilterV2` don't work. Can you explain a bit more?

Reply:
Sorry，这里我理解有误，现在修改为：
```cpp
    if (s.ok()) {
      auto decision = origin_filter_->FilterV2(
          level, key, ValueType::kValue, record.value, new_value, skip_until);
      // It would be a problem if you change the value whereas the value_type is
      // still kBlobIndex. For now, we can just assert value_change == false and
      // abort.
      // TODO: we should make rocksdb Filter API support changing value_type
      assert(decision != CompactionFilter::Decision::kChangeValue);
      return decision;
    }
```

考虑到将来rocksdb的API可能会修改V2的逻辑，因此直接调用V2接口比较合适。